### PR TITLE
[layer] refactor inputs/outputs for layer

### PR DESCRIPTION
**Pull request by kparichay**
Refactor layer to properly handle the number of inputs
and outputs, which is maintained across the in/out dimensions,
in/out data.
This is not maintained for in/out layers as these properties will move
out of layers in the next commit.

Also added bugfix in output_layer which used to clear the output dimension
rendering the num outputs 0 for sometime.

NetworkGraph setNumNetBufferSize is also removed as the above
refactoring handles it at the layer level.

See Also #986

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 1]
```diff
@@ -2143,11 +2143,11 @@ TEST_F(nntrainer_AdditionLayer, initialize_01_p) {
   EXPECT_EQ(status, ML_ERROR_NONE);
 }
 
-TEST_F(nntrainer_AdditionLayer, initialize_02_n) {
+TEST_F(nntrainer_AdditionLayer, initialize_02_p) {
   nntrainer::AdditionLayer layer;
   layer.setProperty({"input_shape=1:1:1:1"});
   status = layer.initialize(manager);
-  EXPECT_EQ(status, ML_ERROR_INVALID_PARAMETER);
+  EXPECT_EQ(status, ML_ERROR_NONE);
```
Comment (zhoonit) [inline test/unittest/unittest_nntrainer_layers.cpp:2142]: Wonder what this was meant to test :p

**Review by zhoonit — APPROVED** [review]
A bit step forward 👍

File: test/unittest/unittest_nntrainer_layers.cpp (hunk 1 above)
Comment (kparichay) [reply test/unittest/unittest_nntrainer_layers.cpp:2142]: I wonder the same.

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**